client/src/controllers/PreviewController.ts

```diff
@@ -0,0 +1,772 @@
+  );
+  const checksPanel = document.querySelector<HTMLElement>(
+    '[data-checks-panel]',
+  );
+  const config = getAxeConfiguration(document.body);
+  const toggleCounter = document.querySelector<HTMLElement>(
+    '[data-side-panel-toggle="checks"] [data-side-panel-toggle-counter]',
+  );
+  const panelCounter = document.querySelector<HTMLElement>(
+    '[data-side-panel="checks"] [data-a11y-result-count]',
+  );
+
+  if (
+    !checksPanel ||
+    !a11yRowTemplate ||
+    !config ||
+    !toggleCounter ||
+    !panelCounter
+  ) {
+    return;
+  }
+
+  // Ensure we only test within the preview iframe, but nonetheless with the correct selectors.
+  config.context = {
+    include: {
+      fromFrames: ['#w-preview-iframe'].concat(
+        (config.context as ContextObject).include as string[],
+      ),
+    },
+  } as ContextObject;
+  if ((config.context.exclude as string[])?.length > 0) {
+    config.context.exclude = {
+      fromFrames: ['#w-preview-iframe'].concat(
+        config.context.exclude as string[],
+      ),
+    } as ContextObject['exclude'];
+  }
+
+  const { results, a11yErrorsNumber } = await getA11yReport(config);
+
+  toggleCounter.innerText = a11yErrorsNumber.toString();
+  toggleCounter.hidden = a11yErrorsNumber === 0;
+  panelCounter.innerText = a11yErrorsNumber.toString();
+  panelCounter.classList.toggle('has-errors', a11yErrorsNumber > 0);
+
+  renderA11yResults(
+    checksPanel,
+    results,
+    config,
+    a11yRowTemplate,
+    onClickSelector,
+  );
+};
+
+interface PreviewDataResponse {
+  is_valid: boolean;
+  is_available: boolean;
+}
+
+/**
+ * Controls the preview panel component to submit the current form state and
+ * update the preview iframe if the form is valid.
+ *
+ * Dispatches the following events, in order:
+ * - `update` before sending the preview data to the server. Cancelable.
+ * - `json` after the preview data update request is completed. Note that the
+ *   data may not be valid and the iframe is not reloaded yet at this point.
+ *   Event `detail`:
+ *   - `data`: the response data that indicates whether the submitted data was
+ *     valid and whether the preview is available.
+ * - `error` when an error occurs while updating the preview data.
+ *   Event `detail`:
+ *   - `error`: the error object that was thrown.
+ * - `load` before reloading the preview iframe. Cancelable.
+ * - `loaded` after the preview iframe has been reloaded.
+ * - `ready` when the preview is ready for further updates – only fired on initial load.
+ * - `updated` after an update cycle is finished. This may or may not involve a
+ *   refresh of the iframe (e.g. in the case of an invalid form).
+ */
+export class PreviewController extends Controller<HTMLElement> {
+  static classes = ['hasErrors', 'selectedSize'];
+
+  static targets = ['size', 'newTab', 'spinner', 'mode', 'iframe'];
+
+  static values = {
+    url: { default: '', type: String },
+    renderUrl: { default: '', type: String },
```

Comment:
This adds the ability to have different URLs for updating the preview data vs the URL for rendering, which can be useful for headless setups. Note that we'll still need to add a way to pass this from the Python code for this feature to be actually useful, but I think that's out of scope for this PR. We can create a new ticket for that after this is merged.

I was thinking about using the `url` value for the rendered preview url and using `dataUrl` value for sending the preview data instead. However, by default we've always used a single URL for both things, and people are more likely to override the rendering URL rather than the data URL. So I think it makes sense to preserve the simpler and straightforward "url" value name for the default stuff (that would end up being used for both values by default), and use the longer `renderUrl` for the one that's likely to be customised.